File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.cpp
class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        int n = arr.size();
        vector<int> dp(k);
        
        for(int i=1;i<=n;++i){
            int highest = 0;
            int currMax = 0;
            for(int j=1;j<=k && i-j>=0;++j){
                currMax = max(currMax, arr[i-j]);
                highest = max(highest,dp[(i-j)%k]+currMax*j);
            }
            dp[i%k] = highest;
        }
        return dp[n%k];
    }
};

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        n = len(arr)
        dp = [0]*k
        
        for i in range(1,n+1):
            highest = currMax = 0
            for j in range(1,min(k,i)+1):
                currMax = max(currMax,arr[i-j])
                highest = max(highest, dp[(i-j)%k]+currMax*j)
            dp[i%k] = highest
        return dp[n%k]